src/radical/pilot/pmgr/launching/base.py

```diff
@@ -595,26 +608,35 @@ def _start_pilot_bulk(self, resource, schema, pilots):
             self._prof.prof('submission_start', uid=pilot['uid'])
 
         # launcher handles pilot job submission
-        launcher = self.find_launcher(rcfg, pilots)
-        launcher.launch_pilots(rcfg, pilots)
+        self.find_launchers(rcfg, pilots)
 
-        for pilot in pilots:
-            self._prof.prof('submission_stop', uid=pilot['uid'])
+        for lname,launcher in self._launchers.items():
+            bucket = list()
+            for pilot in pilots:
+                if pilot['launcher'] == lname:
+                    bucket.append(pilot)
+
+            if bucket:
+                launcher.launch_pilots(rcfg, bucket)
+                for pilot in bucket:
+                    self._prof.prof('submission_stop', uid=pilot['uid'])
 
 
     # --------------------------------------------------------------------------
     #
-    def find_launcher(self, rcfg, pilots):
+    def find_launchers(self, rcfg, pilots):
 
-        for launcher in self._launchers:
-            if launcher.can_launch(rcfg, pilots):
-                self._log.info('use launcher %s for pilots %s',
-                               launcher.name, [p['uid'] for p in pilots])
-                return launcher
-
-        self._log.warn('rcfg: %s', pprint.pformat(rcfg))
-        self._log.warn('pilots: %s', pprint.pformat(pilots))
-        raise RuntimeError('no launcher found for pilots')
+        for pilot in pilots:
+            for lname,launcher in self._launchers.items():
+                if launcher.can_launch(rcfg, pilots):
+                    pilot['launcher'] = lname
+                    self._log.info('use launcher %s for pilot %s',
+                                   lname, pilot['uid'])
+                    return launcher
```

Comment:
- maybe be change the method name to `set_pilot_launchers`?
   - no need to return `launcher` since it is not used
- no need to provide `rcfg`, that information is already in each pilot as `pilot['cfg']['resource_cfg']` after `_prepare_pilot` is called

Reply:
> no need to return launcher since it is not used
Right, the return was wrong.

> maybe be change the method name to set_pilot_launchers?
Yeah, I did not like that either.  I moved that code now into the calling method as it is so small now anyway...